privval/signer_listener_endpoint.go

```diff
@@ -0,0 +1,285 @@
+package privval
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+// SignerValidatorEndpointOption sets an optional parameter on the SocketVal.
+type SignerValidatorEndpointOption func(*SignerListenerEndpoint)
+
+// SignerListenerEndpoint listens for an external process to dial in
+// and keeps the connection alive by dropping and reconnecting
+type SignerListenerEndpoint struct {
+	cmn.BaseService
+
+	mtx      sync.Mutex
+	listener net.Listener
+	conn     net.Conn
+
+	timeoutAccept    time.Duration
+	timeoutReadWrite time.Duration
+
+	stopCh, stoppedCh chan struct{}
+	connectCh         chan struct{}
+	connectedCh       chan net.Conn
+}
+
+// NewSignerListenerEndpoint returns an instance of SignerListenerEndpoint.
+func NewSignerListenerEndpoint(logger log.Logger, listener net.Listener) *SignerListenerEndpoint {
+
+	sc := &SignerListenerEndpoint{
+		listener:         listener,
+		timeoutAccept:    defaultTimeoutAcceptSeconds * time.Second,
+		timeoutReadWrite: defaultTimeoutReadWriteSeconds * time.Second,
+	}
+
+	sc.BaseService = *cmn.NewBaseService(logger, "SignerListenerEndpoint", sc)
+
+	return sc
+}
+
+// OnStart implements cmn.Service.
+func (sl *SignerListenerEndpoint) OnStart() error {
+	sl.Logger.Debug("SignerListenerEndpoint: OnStart")
+
+	sl.stopCh = make(chan struct{})
+	sl.stoppedCh = make(chan struct{})
+
+	sl.connectCh = make(chan struct{})
+	sl.connectedCh = make(chan net.Conn)
+
+	go sl.serviceLoop()
+	sl.connectCh <- struct{}{}
+
+	return nil
+}
+
+// OnStop implements cmn.Service
+func (sl *SignerListenerEndpoint) OnStop() {
+	sl.Logger.Debug("SignerListenerEndpoint: OnStop calling Close")
```

Comment:
I don't think I placed debug statements in repetitive areas that may result in actual slowdowns. I will check the PR again just in case.

In this particular case, calls to OnStop are very sporadic so I expect any effect to be negligible.
Still, I don't mind removing this log entry.